src/nvim/screen.c

```diff
@@ -6017,14 +6047,63 @@ void screenalloc(bool doclear)
   }
 }
 
-void free_screenlines(void)
+void grid_alloc(ScreenGrid *grid, int rows, int columns, bool copy)
 {
-  xfree(ScreenLines);
-  xfree(ScreenAttrs);
-  xfree(LineOffset);
-  xfree(LineWraps);
-  clear_tab_page_click_defs(tab_page_click_defs, tab_page_click_defs_size);
-  xfree(tab_page_click_defs);
+  int new_row, old_row;
+  ScreenGrid new = *grid;
+
+  size_t ncells = (size_t)((rows+1) * columns);
+  new.ScreenLines = xmalloc(ncells * sizeof(schar_T));
+  new.ScreenAttrs = xmalloc(ncells * sizeof(sattr_T));
+  new.LineOffset = xmalloc((size_t)(rows * sizeof(unsigned)));
+  new.LineWraps = xmalloc((size_t)(rows * sizeof(char_u)));
+
+  new.Rows = rows;
+  new.Columns = columns;
+
+  for (new_row = 0; new_row < new.Rows; new_row++) {
+    new.LineOffset[new_row] = new_row * new.Columns;
+    new.LineWraps[new_row] = false;
+    if (copy) {
+      // If the screen is not going to be cleared, copy as much as
+      // possible from the old screen to the new one and clear the rest
+      // (used when resizing the window at the "--more--" prompt or when
+      // executing an external command, for the GUI).
+      memset(new.ScreenLines + new_row * new.Columns,
+             ' ', (size_t)new.Columns * sizeof(schar_T));
+      memset(new.ScreenAttrs + new_row * new.Columns,
+             0, (size_t)new.Columns * sizeof(sattr_T));
+      old_row = new_row + (grid->Rows - new.Rows);
+      if (old_row >= 0 && grid->ScreenLines != NULL) {
+        int len = MIN(grid->Columns, new.Columns);
+        memmove(new.ScreenLines + new.LineOffset[new_row],
+                grid->ScreenLines + grid->LineOffset[old_row],
+                (size_t)len * sizeof(schar_T));
+        memmove(new.ScreenAttrs + new.LineOffset[new_row],
+                grid->ScreenAttrs + grid->LineOffset[old_row],
+                (size_t)len * sizeof(sattr_T));
+      }
+    }
+  }
+  free_screengrid(grid);
+  *grid = new;
+}
+
+
+void free_screengrid(ScreenGrid *grid)
+{
+  xfree(grid->ScreenLines);
+  xfree(grid->ScreenAttrs);
+  xfree(grid->LineOffset);
+  xfree(grid->LineWraps);
+}
+
+void set_screengrid(ScreenGrid *grid)
+{
+  ScreenLines = grid->ScreenLines;
```

Comment:
are we still using global ScreenLines somewhere? If not (or if it could easily be removed), this function can be eliminated.

Reply:
Not in `screen.c` but in other files. About 10-15 instances. Will change those.